File: apps/android/app/src/test/java/net/auralis/app/features/podcasts/EpisodeProgressTest.kt
package net.auralis.app.features.podcasts

import net.auralis.app.data.model.MediaProgress
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class EpisodeProgressTest {
    private fun progress(
        libraryItemId: String = "item1",
        episodeId: String? = "ep1",
        currentTime: Double = 0.0,
        isFinished: Boolean = false,
    ) = MediaProgress(
        id = "prog1",
        libraryItemId = libraryItemId,
        episodeId = episodeId,
        duration = 100.0,
        currentTime = currentTime,
        progress = currentTime / 100.0,
        isFinished = isFinished,
    )

    @Test
    fun `no record is unplayed`() {
        assertEquals(EpisodeProgressState.UNPLAYED, episodeProgressState(null))
    }

    @Test
    fun `a record with zero currentTime and not finished is unplayed`() {
        assertEquals(EpisodeProgressState.UNPLAYED, episodeProgressState(progress(currentTime = 0.0)))
    }

    @Test
    fun `a record with positive currentTime and not finished is in progress`() {
        assertEquals(EpisodeProgressState.IN_PROGRESS, episodeProgressState(progress(currentTime = 42.0)))
    }

    @Test
    fun `isFinished wins over a stale positive currentTime`() {
        assertEquals(
            EpisodeProgressState.PLAYED,
            episodeProgressState(progress(currentTime = 42.0, isFinished = true)),
        )
    }

    @Test
    fun `isFinished with zero currentTime is still played`() {
        assertEquals(
            EpisodeProgressState.PLAYED,
            episodeProgressState(progress(currentTime = 0.0, isFinished = true)),
        )
    }

    @Test
    fun `findEpisodeProgress matches only the exact item and episode pair`() {
        val all =
            listOf(
                progress(libraryItemId = "item1", episodeId = "ep1"),
                progress(libraryItemId = "item1", episodeId = "ep2"),
                progress(libraryItemId = "item2", episodeId = "ep1"),
            )

        val found = findEpisodeProgress(all, "item1", "ep2")

        assertEquals("ep2", found?.episodeId)
        assertEquals("item1", found?.libraryItemId)
    }

    @Test
    fun `findEpisodeProgress never matches a book's item-level record`() {
        val all = listOf(progress(libraryItemId = "item1", episodeId = null))

        val found = findEpisodeProgress(all, "item1", "ep1")

        assertNull(found)
    }

    @Test
    fun `findEpisodeProgress returns null when nothing matches`() {
        val all = listOf(progress(libraryItemId = "item1", episodeId = "ep1"))

        assertNull(findEpisodeProgress(all, "item1", "ep-missing"))
    }
}

File: apps/web/src/features/podcasts/episodeProgress.test.ts
import { describe, expect, it } from 'vitest';
import { episodeProgressState, findEpisodeProgress } from './episodeProgress.js';
import type { MediaProgress } from '../../api/types.js';

function progress(overrides: Partial<MediaProgress>): MediaProgress {
  return {
    id: 'p1',
    libraryItemId: 'item-dailytech',
    episodeId: 'ep-1',
    duration: 300,
    currentTime: 0,
    progress: 0,
    isFinished: false,
    ...overrides,
  };
}

describe('episodeProgressState', () => {
  it('is "unplayed" when there is no progress record at all', () => {
    expect(episodeProgressState(undefined)).toBe('unplayed');
  });

  it('is "unplayed" when a record exists but nothing has played yet', () => {
    expect(episodeProgressState(progress({ currentTime: 0, isFinished: false }))).toBe('unplayed');
  });

  it('is "in-progress" once some time has played but it is not finished', () => {
    expect(episodeProgressState(progress({ currentTime: 42, isFinished: false }))).toBe(
      'in-progress',
    );
  });

  it('is "played" once Audiobookshelf has marked it finished, even if currentTime is stale', () => {
    expect(episodeProgressState(progress({ currentTime: 0, isFinished: true }))).toBe('played');
  });

  it('prefers "played" over "in-progress" when both would otherwise apply', () => {
    expect(episodeProgressState(progress({ currentTime: 42, isFinished: true }))).toBe('played');
  });
});

describe('findEpisodeProgress', () => {
  const records: MediaProgress[] = [
    progress({ id: 'p1', libraryItemId: 'item-dailytech', episodeId: 'ep-1', currentTime: 10 }),
    progress({ id: 'p2', libraryItemId: 'item-dailytech', episodeId: 'ep-2', currentTime: 20 }),
    // A book's progress record: no episodeId. Must never match an episode lookup.
    progress({ id: 'p3', libraryItemId: 'item-dune', episodeId: null, currentTime: 30 }),
  ];

  it('finds the record matching both the item and the episode', () => {
    const found = findEpisodeProgress(records, 'item-dailytech', 'ep-2');
    expect(found?.id).toBe('p2');
  });

  it('returns undefined when no record matches', () => {
    expect(findEpisodeProgress(records, 'item-dailytech', 'ep-3')).toBeUndefined();
  });

  it('never matches a book-level record (null episodeId) against an episode lookup', () => {
    expect(findEpisodeProgress(records, 'item-dune', 'ep-1')).toBeUndefined();
  });
});
